src/freertos_drivers/esp32/Esp32WiFiManager.cxx

```diff
@@ -0,0 +1,614 @@
+    "SSID scan completed",  // WL_SCAN_COMPLETED
+    "WiFi connected",       // WL_CONNECTED
+    "SSID connect failed",  // WL_CONNECT_FAILED
+    "WiFi connection lost", // WL_CONNECTION_LOST
+    "WiFi disconnected"     // WL_DISCONNECTED
+};
+
+Esp32WiFiManager::Esp32WiFiManager(const char *ssid, const char *password,
+    const char *hostname_prefix, OpenMRN *openmrn,
+    const openlcb::NodeID node_id, const WiFiConfiguration &cfg) : ssid_(ssid),
+    password_(password), cfg_(cfg), openmrn_(openmrn)
+{
+    // register the Esp32WiFiManager instance to receive config updates.
+    ConfigUpdateService::instance()->register_update_listener(this);
+
+    // if these are not passed in we will not manage the WiFi stack with this
+    // class, only the hub and outbound connection features.
+    if (ssid_ && password_ && hostname_prefix)
+    {
+        // calculate the hostname for the ESP32 based on the provided prefix and
+        // the last 32 bits of the node id.
+        // node_id : 0x050101011425
+        // hostname_ : hostname_prefix + '-' + 01011425
+        hostname_ = String(hostname_prefix) + "-";
+        // append the last 32 bits of the provided node id, ie:
+        hostname_ += String((uint32_t)(node_id & 0xFFFFFFFF), 16);
+
+        // the maximum length hostname for the ESP32 is 32 bytes so truncate if
+        // necessary
+        // ref https://github.com/espressif/esp-idf/blob/master/components/tcpip_adapter/include/tcpip_adapter.h#L611
+        if (hostname_.length() > TCPIP_HOSTNAME_MAX_SIZE)
+        {
+            LOG(WARNING, "ESP32 hostname is too long, original hostname: %s", hostname_.c_str());
+            hostname_ = hostname_.substring(0, TCPIP_HOSTNAME_MAX_SIZE);
+            LOG(WARNING, "truncated hostname: %s", hostname_.c_str());
+        }
+        manageWiFi_ = true;
+    }
+    else
+    {
+        // since we do not have an SSID, Password and hostname prefix we should
+        // not manage the WiFi stack within this class.
+        manageWiFi_ = false;
+    }
+}
+
+#define VERIFY_CACHED_VALUE(current_value, new_value) \
+    if (current_value != new_value) \
+    { \
+        current_value = new_value; \
+        configUpdated = true; \
+    }
+
+ConfigUpdateListener::UpdateAction Esp32WiFiManager::apply_configuration(
+    int fd, bool initial_load, BarrierNotifiable *done)
+{
+    AutoNotify n(done);
+    LOG(VERBOSE, "Esp32WiFiManager::apply_configuration(%d, %d)", fd, initial_load);
+    bool configUpdated = false;
+    bool wifiSleepMode = cfg_.wifi_sleep().read(fd);
+    bool hubEnabled = cfg_.hub_config().enabled().read(fd);
+    uint16_t hubPort = cfg_.hub_config().port().read(fd);
+    String hubServiceName = String(cfg_.hub_config().service_name().read(fd).c_str());
+    const TcpClientConfig<TcpClientDefaultParams> &link_config =
+        cfg_.link_config();
+    SocketClientParams::SearchMode tcpClientMode =
+        (SocketClientParams::SearchMode)link_config.search_mode().read(fd);
```

Comment:
I would suggest using CDI_READ_TRIMMED(...) macro for getting boolean or other numeric values from the CDI, as it will ensure that the result falls within range of valid values.